core/toolbox.js

```diff
@@ -252,6 +252,156 @@ Blockly.Toolbox.prototype.renderTree = function(languageTree) {
+ *    to the recursive nature of this function, treeOut can be either the root of
+ *    the tree (Blockly.tree.TreeControl) or a child node of the tree
+ *    (Blockly.tree.TreeNode). These nodes are built from the toolboxDef.
+ * @return {Blockly.tree.BaseNode} The TreeNode to expand when the toolbox is
+ *    first loaded (or null).
+ * @private
+ */
+Blockly.Toolbox.prototype.createTree_ = function(toolboxDef, treeOut) {
+  var openNode = null;
+  var lastElement = null;
+  if (!toolboxDef) {
+    return null;
+  }
+
+  for (var i = 0, childIn; (childIn = toolboxDef[i]); i++) {
+    switch (childIn.kind.toUpperCase()) {
+      case 'CATEGORY':
+        var categoryInfo = /** @type {Blockly.utils.toolbox.Category} */ (childIn);
+        openNode = this.addCategory_(categoryInfo, treeOut) || openNode;
+        lastElement = childIn;
+        break;
+      case 'SEP':
+        var separatorInfo = /** @type {Blockly.utils.toolbox.Separator} */ (childIn);
+        lastElement = this.addSeparator_(separatorInfo, treeOut, lastElement) || lastElement;
+        break;
+      case 'BLOCK':
+      case 'SHADOW':
+      case 'LABEL':
+      case 'BUTTON':
+        treeOut.contents.push(childIn);
+        lastElement = childIn;
+        break;
+    }
+  }
+  return openNode;
+};
+
+/**
+ * Add a category to the toolbox tree.
+ * @param {!Blockly.utils.toolbox.Category} categoryInfo The object holding
+ *    information on the category.
+ * @param {!Blockly.tree.BaseNode} treeOut The TreeControl or TreeNode
+ *     object built from the childNodes.
+ * @return {Blockly.tree.BaseNode} TreeNode to open at startup (or null).
+ * @private
+ */
+Blockly.Toolbox.prototype.addCategory_ = function(categoryInfo, treeOut) {
+  var openNode = null;
+  // Decode the category name for any potential message references
+  // (eg. `%{BKY_CATEGORY_NAME_LOGIC}`).
+  var categoryName = Blockly.utils.replaceMessageReferences(categoryInfo['name']);
+
+  // Create and add the tree node for the category.
+  var childOut = this.tree_.createNode(categoryName);
+  childOut.onSizeChanged(this.handleNodeSizeChanged_);
+  childOut.contents = [];
+  treeOut.add(childOut);
+
+  var custom = categoryInfo['custom'];
+
+  if (custom) {
+    // Variables and procedures are special dynamic categories.
+    childOut.contents = custom;
+  } else {
+    openNode = this.createTree_(categoryInfo.contents, childOut) || openNode;
+  }
+  this.addColourOrStyle_(categoryInfo, childOut, categoryName);
+  openNode = this.setExpanded_(categoryInfo, childOut) || openNode;
+  return openNode;
+};
+
+/**
+ * Add either the colour or the style for a category.
+ * @param {!Blockly.utils.toolbox.Category} categoryInfo The object holding
+ *    information on the category.
+ * @param {!Blockly.tree.TreeNode} childOut The TreeNode for a category.
+ * @param {string} categoryName The name of the category.
+ * @private
+ */
+Blockly.Toolbox.prototype.addColourOrStyle_ = function(
```

Comment:
Are you adding or setting the colour here?
For the sake of consistency with the methods below, should this be setColourOrStyle_ ?

Reply:
done